_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,791 @@
+
+```rust
+use askama::Template;
+use askama_axum::IntoResponse as AskamaIntoResponse;
+use axum::{
+    extract::{Form, Path, State},
+    http::StatusCode,
+    response::{IntoResponse as AxumIntoResponse, Redirect, Response},
+    routing::{get, post},
+    Router,
+};
+use chrono::{DateTime, Utc};
+use futures_util::StreamExt;
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+use sqlx::PgPool;
+use tokio::time::{sleep, Duration};
+use validator::Validate;
+```
+
+## Building
+
+There are three main parts to this: The frontend, the backend, and the actual monitoring task itself. Let’s start with the monitoring task first.
+
+### Monitoring task
+
+The monitoring task itself is simple enough: we fetch the list of websites from the database, then sequentially send a HTTP request to each of them and record the results in Postgres. Firstly, we’ll create the struct that we want to represent the `Website`:
+
+```rust
+#[derive(Deserialize, sqlx::FromRow, Validate)]
+struct Website {
+    #[validate(url)]
+    url: String,
+    alias: String
+}
+```
+
+Here we instantiate a `reqwest` client and then fetch all of the websites that we want to search for:
+
+```rust
+async fn check_websites(db: PgPool) {
+    let ctx = Client::new();
+
+    let mut res = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites").fetch(&db);
+
+    // .. rest of your code
+}
+```
+
+ We *could* use `fetch_all()`, but by using `fetch()` we get a `Stream` back and can skip dealing with `RowNotFound` SQL errors by checking whether or not there are any rows to fetch. 
+
+Now that we have our list of websites (or lack thereof), we can try to send a request sequentially to each website that exists then store the results of the fetch in our database:
+
+```rust
+while let Some(website) = res.next().await {
+    let website = website.unwrap();
+
+    let response = ctx.get(website.url).send().await.unwrap();
+
+    sqlx::query(
+        "INSERT INTO logs (website_alias, status)
+            VALUES
+            ($1, $2)"
+    )
+        .bind(website.alias)
+        .bind(response.status().as_u16() as i16)
+        .execute(&db).await
+        .unwrap();
+}
+```
+
+While we don’t store the response body, at this point there’s not much reason to do so. The response status should give us all the information we need.
+
+After that, we then need to figure out when we want to run the request loop again. Here we have a snippet that calculates the time left until two seconds into the next minute, then waits until the requested time:
+
+```rust
+let next_time = Utc::now().format("%Y/%m/%d %H:%M:02.000 %z").to_string();
+let next_time =
+    DateTime::parse_from_str(&next_time, "%Y/%m/%d %H:%M:%S%.3f %z").unwrap() +
+    chrono::Duration::seconds(60);
+
+let duration_to_wait = next_time.signed_duration_since(Utc::now()).num_seconds() as u64;
+
+sleep(Duration::from_secs(duration_to_wait)).await;
```

Comment:
It seems like you want this to run every minute. Why not use the [`tokio::time::interval`](https://docs.rs/tokio/latest/tokio/time/fn.interval.html) instead?